example/timer/odp_timer_accuracy.c

```diff
@@ -0,0 +1,461 @@
+
+	while (num_left) {
+		ev = odp_schedule(NULL, ODP_SCHED_WAIT);
+
+		time = odp_time_local();
+		time_ns = odp_time_to_ns(time);
+
+		if (time_ns > next_tmo) {
+			diff_ns = time_ns - next_tmo;
+			num_after++;
+			after += diff_ns;
+			if (diff_ns < min_after)
+				min_after = diff_ns;
+			if (diff_ns > max_after)
+				max_after = diff_ns;
+
+		} else if (time_ns < next_tmo) {
+			diff_ns = next_tmo - time_ns;
+			num_before++;
+			before += diff_ns;
+			if (diff_ns < min_before)
+				min_before = diff_ns;
+			if (diff_ns > max_before)
+				max_before = diff_ns;
+
+		} else {
+			num_exact++;
+		}
+
+		odp_event_free(ev);
+
+		next_tmo += test_global->period_ns;
+		num_left--;
+	}
+
+	/* Free current scheduler context. There should be no more events. */
+	while ((ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT))
+	       != ODP_EVENT_INVALID) {
+		printf("Dropping extra event\n");
+		odp_event_free(ev);
+	}
+
+	if (num_after)
+		ave_after = (double)after / num_after;
+	else
+		min_after = 0;
+
+	if (num_before)
+		ave_before = (double)before / num_before;
+	else
+		min_before = 0;
+
+	printf("\n Test results:\n");
+	printf("  num after:  %12i  /  %.2f%%\n",
+	       num_after, 100.0 * num_after / num);
+	printf("  num before: %12i  /  %.2f%%\n",
+	       num_before, 100.0 * num_before / num);
+	printf("  num exact:  %12i  /  %.2f%%\n",
+	       num_exact, 100.0 * num_exact / num);
+	printf("  error after (nsec):\n");
+	printf("         min: %12" PRIu64 "  /  %.3fx resolution\n",
+	       min_after, (double)min_after / res_ns);
+	printf("         max: %12" PRIu64 "  /  %.3fx resolution\n",
+	       max_after, (double)max_after / res_ns);
+	printf("         ave: %12.0f  /  %.3fx resolution\n",
+	       ave_after, ave_after / res_ns);
+	printf("  error before (nsec):\n");
+	printf("         min: %12" PRIu64 "  /  %.3fx resolution\n",
+	       min_before, (double)min_before / res_ns);
+	printf("         max: %12" PRIu64 "  /  %.3fx resolution\n",
+	       max_before, (double)max_before / res_ns);
+	printf("         ave: %12.0f  /  %.3fx resolution\n",
+	       ave_before, ave_before / res_ns);
+	printf("\n");
+}
+
+int main(int argc, char *argv[])
+{
+	odp_instance_t instance;
+	odp_init_t init;
+	int num;
+	test_global_t test_global;
+	odp_init_t *init_ptr = NULL;
+	int ret = 0;
+
+	memset(&test_global, 0, sizeof(test_global_t));
+
+	if (parse_options(argc, argv, &test_global))
+		return -1;
+
+	/* List features not to be used (may optimize performance) */
+	odp_init_param_init(&init);
+	init.not_used.feat.cls    = 1;
+	init.not_used.feat.crypto = 1;
+	init.not_used.feat.ipsec  = 1;
+	init.not_used.feat.tm     = 1;
+
+	if (test_global.opt.init)
+		init_ptr = &init;
+
+	/* Init ODP before calling anything else */
+	if (odp_init_global(&instance, init_ptr, NULL)) {
+		printf("Global init failed.\n");
+		return -1;
+	}
+
+	/* Init this thread */
+	if (odp_init_local(instance, ODP_THREAD_CONTROL)) {
+		printf("Local init failed.\n");
+		return -1;
```

Comment:
goto 455

Reply:
In practice global/local init failure is fatal. I'd not complicate this simple app with code that tries to survive init failure. Print and exit is enough.